homeassistant/components/stream/recorder.py

```diff
@@ -135,14 +42,115 @@ def prepend(self, segments: list[Segment]) -> None:
+            if not output:
+                output = av.open(
+                    self.video_path + ".tmp",
+                    "w",
+                    format=RECORDER_CONTAINER_FORMAT,
+                    container_options={
+                        "video_track_timescale": str(int(1 / source_v.time_base))
+                    },
+                )
+
+            # Add output streams if necessary
+            if not output_v:
+                output_v = output.add_stream(template=source_v)
+                context = output_v.codec_context
+                context.flags |= "GLOBAL_HEADER"
+            if source_a and not output_a:
+                output_a = output.add_stream(template=source_a)
+
+            # Recalculate pts adjustments on first segment and on any discontinuity
+            # We are assuming time base is the same across all discontinuities
+            if last_stream_id != segment.stream_id:
+                last_stream_id = segment.stream_id
+                pts_adjuster["video"] = int(
+                    (running_duration - source.start_time)
+                    / (av.time_base * source_v.time_base)
+                )
+                if source_a:
+                    pts_adjuster["audio"] = int(
+                        (running_duration - source.start_time)
+                        / (av.time_base * source_a.time_base)
+                    )
+
+            # Remux video
+            for packet in source.demux():
+                if packet.dts is None:
+                    continue
+                packet.pts += pts_adjuster[packet.stream.type]
+                packet.dts += pts_adjuster[packet.stream.type]
+                packet.stream = output_v if packet.stream.type == "video" else output_a
+                output.mux(packet)
+
+            running_duration += source.duration - source.start_time
+
+            source.close()
+
+        # Write lookback segments
+        while len(self._segments) > 1:  # The last segment is in progress
+            await self._hass.async_add_executor_job(
+                write_segment, self._segments.popleft()
+            )
+        # Make sure the first segment has been added
+        if not self._segments:
+            await self.recv()
+        # Write segments as soon as they are completed
+        while not self.idle:
+            await self.recv()
+            await self._hass.async_add_executor_job(
+                write_segment, self._segments.popleft()
+            )
+        # Write remaining segments
+        # Should only have 0 or 1 segments, but loop through just in case
+        while self._segments:
+            await self._hass.async_add_executor_job(
+                write_segment, self._segments.popleft()
+            )
+        if output is None:
+            _LOGGER.error("Recording failed to capture anything")
+        else:
+            output.close()
+            os.rename(self.video_path + ".tmp", self.video_path)
```

Comment:
I put this part in (writing to a file with a different name first) because I believe some users currently use the existence of the output file to check whether the recording is done yet.